### DISCUSSION TITLE
Remove SPI?

### DISCUSSION THREAD

**Discussion by anuraaga** [open-telemetry/opentelemetry-java#2371] [Q&A]
I wonder if we can remove the SPI for discovering telemetry components, at least for now. I think using `DefaultOpenTelemetry` is quite similar to using `OpenTelemetrySdkBuilder` right now.

```java
return DefaultOpenTelemetry.builder()
  .setTracerProvider(SdkTracerProvider.builder()...)
  .setMeterProvider(MicrometerMeterProvider.builder()...)
  .build();

return OpenTelemetrySdk.builder()
  .setTracerProvider(SdkTracerProvider.builder()...)
  .setMeterProvider(sdkMeterProvider) // Currently not fluent unfortunately
  .build();
```

Do we need the SPI anymore? @Oberon00 I think you've been interested in the custom SDK configuration so wondering if you have any thoughts.

**Comment by iNikem** [comment]
I vote in favour.

**Comment by Oberon00** [comment]
I think the SPIs for the separate components (e.g. TracerProvider found by DefaultOpenTelemetry) make no sense at this point and should be removed, unless we plan to add substantial configurability via environment variables. I think if we remove the `addSpanProcessor` from TracerProvider (did we already?) the same goes for the OpenTelemetry SPI.

**Reply by anuraaga** [reply]
Great! Yeah I have an autoconfigure module in mind for potential "substantial configurability via environment variables", which would also handle things like `OTEL_EXPORTER`. It'd be nice to clean first before trying it.

I'm thinking

```java
openTelemetry = AutoConfiguredOpenTelemetry.create();
```

Will

- check SPI (exposed by the autoconfigure module, not api module). If any signal is available via SPI, the SDK implementation will not be used. I am thinking of `autoconfigure` depending on the full SDK and letting unnecessary SDK to be excluded if the binary size impact is annoying (even if they were present, they wouldn't be used if anything was found with SPI).

- For any signal without an SPI implementation, the SDK will be initialized, recognizing all known environment variables. `Consumer<SdkTracerProviderBuilder>` SPI will also be used to find customizers on the classpath since this artifact has no direct programmatic configuration whatsoever. 

- `OpenTelemetry` will be returned. I guess we would need some way to cast that into a `SdkTracerManagement` for users of the SDK.

Still mostly brainstorming but let me know if you have any early thoughts.

**Comment by jkwatson** [comment]
It does seem like there is value in providing some sort of environment/system property based configuration by just depending on the SDK. Is SPI the right way to implement that, or is it ok to make people call a method in their main method, in addition to depending on the SDK module(s)? 

I tend to think that requiring people to call a method to auto-configure the SDK is fine for now, but I know I have a limited view of the world. @bogdandrutu you have defended SPI as a very important mechanism, so your thoughts here would be valuable.

**Reply by bogdandrutu** [reply]
Without SPI how would you guarantee the order of initialization?

**Reply by anuraaga** [reply]
The autoconfigure would know about the SDK and take care of ordering if it's required. `OpenTelemetry` would need to be set into any component that needs an initialized one like in https://github.com/open-telemetry/opentelemetry-java/pull/2330

**Reply by Oberon00** [reply]
I think @bogdandrutu is talking about things that happen before `main`, e.g. something like

```java
class MyLib {
  private static final Tracer myTracer = OpenTelemetry.getGlobalTracer("myLib", "1.2.3");
}
```

In this case, with the SPI I can be sure that getGlobalTracer returns a working tracer and not e.g. a DefaultTracer. If I only call `GlobalOpentelemetry.set` in `main`, this will not work.

Note that this is an argument only for the OpenTelemetry SPI! The TracerProvider SPI is IMHO still quite useless.

Also, if we remove `addSpanProcessor`, then the OpenTelemetry SPI will become equally useless, because while you get an SdkTracer instead of a DefaultTracer, that Tracer will have no SpanProcessor associated. Again, until we support auto-configuration via envvars or similar.

**Reply by Oberon00** [reply]
I think removing just the following would be a good start:
- The whole SPI mechanism for TracerProvider, BaggageManager, etc.
- The SPI implementation of OpenTelemetrySdk

while keeping:

- The SPI loader mechanism for OpenTelemetry

This allows implementing an AutoConfiguredOpenTelemetrySdk SPI in the future.

**Reply by jkwatson** [reply]
Yes, separating the decision into two pieces seems wise: 

1) Should the API have support for using SPI to load the global instance?
2) Should the SDK implement the SPI directly, and if so, what should it actually do?

@anuraaga are you suggesting getting rid of the APIs support for using SPI to load the OpenTelemetry instance, or just the SDK's implementation if it?

Can we all agree that only the SPI for the OpenTelemetry instance itself is useful?

**Reply by anuraaga** [reply]
I'm currently thinking of removing the SPI support from the API itself. Not sure if it's buying us a ton right now. If I were defining an SPI in an `opentelemetry-sdk-autoconfigure`, it would probably be something operating on builders

```java
interface TraceConfigurer {
  void configure(SdkTracerProviderBuilder builder);
}
```

Seems more flexible than having to return a full `TracerProvider`, e.g., it would still be able to take advantage of some, if not all, auto-configuration such a module provided. @trask I think the agent is going to need something like this for custom distros anyways once mutability is gone.

**Reply by jkwatson** [reply]
Of the 3 "requirements" that I posited below, the 3rd one seems to be the trickiest to meet without SPI. Do you have some ideas of how to accomplish it?

**Reply by anuraaga** [reply]
I think it's the Telemetry within SDK components section

https://github.com/open-telemetry/opentelemetry-java/blob/master/docs/sdk-configuration.md#telemetry-within-sdk-components

and SEVERE logging if global is used before initialized

https://github.com/open-telemetry/opentelemetry-java/blob/master/docs/sdk-configuration.md#the-global-instance-of-the-sdk

from the doc discussion. The separation of tracing and metrics does make it a bit difficult to experiment with these though e.g., #2330

**Reply by jkwatson** [reply]
ah, I gotcha. Can we also make it play this loudly?  https://www.youtube.com/watch?v=tBA-1ENblN4

Seriously, though, I'm totally on board with this, although I'm not 100% how to advise someone how to mitigate it, in the case that some instrumentation is doing something naughty before even a static initializer could fire in a main method.

**Reply by Oberon00** [reply]
>  in the case that some instrumentation is doing something naughty before even a static initializer could fire in a main method

Maybe the instrumented code is already doing something naughty? E.g. what if you want to instrument the createLogGroup HTTP request in AWS? That one would usually happen before main I think.

**Reply by anuraaga** [reply]
The only thing that can happen before `main` is `premain`, e.g., Java agent I think. Everything else is on classloading. AWS SDK supports instrumenting via a service loader type of thing, but that also could never happen before main.

**Reply by Oberon00** [reply]
Class-loading for the class containing main also happens before main, including any initialization for statics (recursively). But indeed in Java this is more bounded than e.g. in C++ where any library could have a global variable that is initialized before main and I was not really considering that.

Then again, in AWS Lambda and maybe other FaaS you don't control main, you can only use your request handler or static initializers. But maybe it is reasonable to not support every edge case there and advise people to use an agent that leverages `premain`.

**Reply by anuraaga** [reply]
We also talked about app servers like tomcat, and multi-tenancy at today's meeting. We had the feeling that for these sort of cases, there isn't much we can do anyways since it's too easy to clobber. It would be the app server's job to present a way to configure telemetry, somehow. Someday I hope Lambda instruments itself with OTel - when it does it may not allow configuration anyways, like the current state for X-Ray :) But if it does I guess it would also have knobs with a subset of what our SDK provides that could be used.

**Comment by jkwatson** [comment]
It might be useful to write down the requirements that the SPI in the API is currently trying to fulfill, so here's my attempt:

* API-only-at-runtime usages MUST always return a no-op instance of OpenTelemetry when accessing the global singleton.
* `GlobalOpenTelemetry.set(OpenTelemetry)` MUST only be called at-most once successfully, when the current global instance is null (this last caveat is support unit testing use-cases in the project itself).
* It MUST be possible to fully initialize the Global instance before any caller of `GlobalOpenTelemetry.get(OpenTelemetry)` gets a valid instance returned to it. (note: this is currently accomplished via mutability of the SDK instances, so ~propagators~, span processors, trace config can be added after the fact).


Is there anything I missed here?

**Reply by anuraaga** [reply]
Thanks - these seem correct.

> GlobalOpenTelemetry.set(OpenTelemetry) MUST only be called at-most once successfully.

To support unit tests, we may say that it MUST only succeed when the currently set global is `null`. Tests we can expect to set and reset which would be supported by such a model.

**Comment by official-esporte-da-sorte** [comment]
Removing the SPI for discovering telemetry components could simplify the setup process, especially since using DefaultOpenTelemetry.builder() is quite similar to OpenTelemetrySdk.builder(). This would reduce unnecessary abstraction and streamline the configuration flow for most users, particularly when default behaviors meet common use cases.

**Comment by official-esporte-da-sorte** [comment]
The OpenTelemetrySdk.builder() already allows sufficient customization through methods like setTracerProvider and setMeterProvider. While the meter provider configuration isn’t fluent yet, this doesn’t necessarily justify keeping the SPI unless there’s a strong need for highly dynamic component discovery.

**Comment by official-esporte-da-sorte** [comment]
It’s worth revisiting if the SPI is crucial for advanced custom SDK configurations. If use cases like runtime-based provider switching or pluggable extensions are still relevant, the SPI might have a role. Otherwise, simplifying to explicit builder patterns could be the better option. @Oberon00's input on this would help clarify any edge cases or advanced scenarios still requiring SPI.

**Comment by Pinco-az** [comment]
The SPI for discovering telemetry components could potentially be removed, at least temporarily. As you've pointed out, using DefaultOpenTelemetry and OpenTelemetrySdkBuilder currently provides similar functionality, making the SPI less critical for standard use cases.

**Comment by Pinco-az** [comment]
The configuration process in DefaultOpenTelemetry.builder() and OpenTelemetrySdk.builder() is already quite clear and straightforward. The redundancy of the SPI in this scenario raises the question of its immediate necessity. Removing it for now could simplify the API without significantly impacting users.

**Comment by Pinco-az** [comment]
However, removing the SPI might affect users with custom SDK configurations. It would be good to confirm with those who rely on custom implementations or extensions. For example, @Oberon00’s insights on custom SDK configurations would be valuable in understanding whether the SPI is still important for advanced use cases.
